hash/hash.go

```diff
@@ -12,32 +13,33 @@ const LabelHashNestingDepth = 100
 //
 // ⚠️ The resulting function is NOT thread-safe, however different generated instances are independent.
 // The code is optimized for performance and memory allocations.
-func GenMerkleHashFunc(challenge []byte) func(lChild, rChild []byte) []byte {
-	var buffer []byte
-	return func(lChild, rChild []byte) []byte {
-		size := len(challenge) + len(lChild) + len(rChild)
-		if len(buffer) < size {
-			buffer = make([]byte, size)
-		}
-		copy(buffer, challenge)
-		copy(buffer[len(challenge):], lChild)
-		copy(buffer[len(challenge)+len(lChild):], rChild)
-
-		result := sha256.Sum256(buffer[:size])
-		return result[:]
+func GenMerkleHashFunc(challenge []byte) merkle.HashFunc {
+	h := sha256.New()
+	return func(buf, lChild, rChild []byte) []byte {
+		h.Reset()
+		_, _ = h.Write(challenge)
+		_, _ = h.Write(lChild)
+		_, _ = h.Write(rChild)
+		return h.Sum(buf)
 	}
 }
 
 // GenLabelHashFunc generates hash functions for computing labels. The challenge is prepended to the data and the result
 // is hashed using Sha256. TODO: use nested hashes based on a difficulty param.
 func GenLabelHashFunc(challenge []byte) func(data []byte) []byte {
+	var hashBuf [sha256.Size]byte
+	h := sha256.New()
 	return func(data []byte) []byte {
-		message := append(challenge, data...)
-		var res [32]byte
-		for i := 0; i < LabelHashNestingDepth; i++ {
-			res = sha256.Sum256(message)
-			message = res[:]
+		h.Reset()
+		_, _ = h.Write(challenge)
+		_, _ = h.Write(data)
+		hash := h.Sum(hashBuf[:0])
```

Comment:
would `h.Sum(nil)` here not have the same effect?

Reply:
The difference is significant:
with `nil`:

> BenchmarkGetProof-20    	       1	30001720023 ns/op	   4.49 MB/s	   4205877 leafs/op	    140196 leafs/sec	14240215608 B/op	420594083 allocs/op

with reusing the buffer:
> BenchmarkGetProof-20    	       1	30001170355 ns/op	   5.57 MB/s	   5218087 leafs/op	    173936 leafs/sec	977524200 B/op	    6333 allocs/op

Reply:
`hashBuf` is an array that gets used as the backing array of a new slice here. It is true that calling the returned closure multiple times saves allocating memory, but I think it would be more efficient for the runtime if `hashBuf` would be a slice instead of an array: `hashBuf := make([]byte, 0, sha256.Size)` not requiring the "cast" from array to slice on every call here.